### PR TITLE
Add new flag -Wreturn-mismatch

### LINKED ISSUE #72116: Modify -Wreturn-type and add -Wreturn-mismatch from GCC 14
GCC 14 modified the way `-Wreturn-type` behaves to split some of its functionality off into `-Wreturn-mismatch` and strengthened some of the diagnostics to be errors by default (with the intention of making them permaerrors in the future). Clang already considers many of these cases to be errors by default already, but we should pick up the new warning flag and try to split our diagnostics similarly for compatibility reasons.

From https://discourse.llvm.org/t/gcc-14-adds-wreturn-mismatch/74868/1

### PR DESCRIPTION
**Pull request by 11happy**
**Overview:**
This pull request fixes #72116 where a new flag is introduced for compatibility with GCC 14, the functionality of -Wreturn-type is modified to split some of its behaviors into -Wreturn-mismatch 

**Testing:**
- Tested the updated code.
- Verified that other functionalities remain unaffected.
Screenshot from 2024-02-24 13-56-24
- Sample working:
```
int foo(int x){
  x = 1;
  return;
}
```
Screenshot from 2024-02-24 14-01-21

**Dependencies:**
- No dependencies on other pull requests.

### CODE REVIEW

**Comment by 11happy** [comment]
Its a Humble Ping!

File: clang/test/Sema/return.c [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %clang_cc1 -triple x86_64-apple-darwin9 %s -fsyntax-only -Wignored-qualifiers -Wno-error=return-type -Wno-error=implicit-int -verify -fblocks -Wno-unreachable-code -Wno-unused-value -Wno-strict-prototypes
+// RUN: %clang_cc1 -triple x86_64-apple-darwin9 %s -fsyntax-only -Wignored-qualifiers -Wno-error=return-type -Wno-error=return-mismatch -Wno-error=implicit-int -verify -fblocks -Wno-unreachable-code -Wno-unused-value -Wno-strict-prototypes
 
 // clang emits the following warning by default.
 // With GCC, -pedantic, -Wreturn-type or -Wall are required to produce the
```
Comment (AaronBallman) [inline clang/test/Sema/return.c:4]: This comment is no longer accurate and should be updated.

File: clang/test/Analysis/null-deref-ps.c [hunk 2]
```diff
@@ -1,5 +1,5 @@
-// RUN: %clang_analyze_cc1 -triple i386-apple-darwin10 -Wno-int-conversion -Wno-strict-prototypes -Wno-tautological-constant-compare -Wtautological-unsigned-zero-compare -analyzer-checker=core,deadcode,alpha.core -std=gnu99 -analyzer-purge=none -verify %s -Wno-error=return-type
-// RUN: %clang_analyze_cc1 -triple i386-apple-darwin10 -Wno-int-conversion -Wno-strict-prototypes -Wno-tautological-constant-compare -Wtautological-unsigned-zero-compare -analyzer-checker=core,deadcode,alpha.core -std=gnu99 -verify %s -Wno-error=return-type
+// RUN: %clang_analyze_cc1 -triple i386-apple-darwin10 -Wno-int-conversion -Wno-strict-prototypes -Wno-tautological-constant-compare -Wtautological-unsigned-zero-compare -analyzer-checker=core,deadcode,alpha.core -std=gnu99 -analyzer-purge=none -verify %s -Wno-error=return-type -Wno-error=return-mismatch
```
Comment (AaronBallman) [inline clang/test/Analysis/null-deref-ps.c:1]: Why is this change needed? I would have assumed the `-Wno-error=return-type` would have been sufficient because `-Wreturn-mismatch` is grouped under it. (The same question applies to all the test changes.)

File: clang/test/CodeGen/volatile-1.c [hunk 3]
```diff
@@ -1,5 +1,5 @@
 // XFAIL: target=aarch64-pc-windows-msvc
-// RUN: %clang_cc1 -Wno-return-type -Wno-unused-value -emit-llvm %s -w -o - | FileCheck %s
+// RUN: %clang_cc1 -Wno-return-type -Wno-return-mismatch -Wno-unused-value -emit-llvm %s -w -o - | FileCheck %s
```
Comment (AaronBallman) [inline clang/test/CodeGen/volatile-1.c:2]: Similarly, why is this change required -- I would have assumed that `-Wno-return-type` was sufficient to silence both. (Same question applies to `return-silent.c`.)

**Review by AaronBallman — COMMENTED** [review]
Thank you for the changes! Please be sure to add a release note to `clang/docs/ReleaseNotes.rst` so users know about the new diagnostic flag, and the changes should come with more test coverage. I'd like to see to see a test with `-Wreturn-type -Wno-return-mismatch` and `-Wno-return-type -Wreturn-mismatch` to demonstrate that we properly enable/disable the correct diagnostics. Be sure to test the test cases against GCC trunk as well to see whether we match GCC's behavior or not.

File: clang/test/Sema/return.c (hunk 1 above)
Comment (11happy) [reply clang/test/Sema/return.c]: done

File: clang/test/Analysis/null-deref-ps.c (hunk 2 above)
Comment (11happy) [reply clang/test/Analysis/null-deref-ps.c]: sorry earlier I created a new seperate return-mismatch & tests were failing corrected it.

File: clang/test/CodeGen/volatile-1.c (hunk 3 above)
Comment (11happy) [reply clang/test/CodeGen/volatile-1.c]: same as above. done.

File: clang/test/Sema/return-mismatch.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+int test4() {
+    (void)(bar() || foo()); // no-warning
+} 
+
+void test5() {
+} // no-warning
+
+int test6() {
+  return 0; // no-warning
+}
```
Comment (AaronBallman) [inline clang/test/Sema/return-mismatch.c:29]: Please add a newline to the end of the file.

File: clang/test/Sema/return-type.c [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// RUN: %clang_cc1 -Wreturn-type -Wno-return-mismatch -fsyntax-only -verify %s
```
Comment (AaronBallman) [inline clang/test/Sema/return-type.c:1]: Because this is the same test code in both files, let's remove this file, add this RUN line to the other file, and rename the other file `return-type-mismatch.h`.

File: clang/test/Sema/return-mismatch.c [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+int foo(void) __attribute__((noreturn));
+int bar(void);
+
+void test1() {
+  return 1; // expected-warning{{void function 'test1' should not return a value}}
+}
+
+int test2() { 
+    return; // expected-warning{{non-void function 'test2' should return a value}}
```
Comment (AaronBallman) [inline clang/test/Sema/return-mismatch.c:13]:
```suggestion
    return; // return-mismatch-warning{{non-void function 'test2' should return a value}}
```

File: clang/test/Sema/return-mismatch.c [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+} 
+
+int test3() { 
+    // no-warning
```
Comment (AaronBallman) [inline clang/test/Sema/return-mismatch.c:17]: When you add the other RUN line, you'll use `// return-type-warning@+1 {{non-void function does not return a value}}` (similar for the other return-type.c warnings that will show up in this file).

File: clang/test/Sema/return-mismatch.c (hunk 4 above)
Comment (AaronBallman) [reply clang/test/Sema/return-mismatch.c:29]: Some other test cases to add:
```
int test7(void) {
  foo(); // no warning
}

int test8(void) {
  bar(); // warning: non-void function does not return a value
}
```

File: clang/test/Sema/return-mismatch.c [hunk 8]
```diff
@@ -0,0 +1,29 @@
+// RUN: %clang_cc1 -Wno-return-type -Wreturn-mismatch -fsyntax-only -verify %s
+
+// Test that -Wreturn-mismatch is enabled and -Wreturn-type is disabled.
+
+int foo(void) __attribute__((noreturn));
+int bar(void);
+
+void test1() {
```
Comment (AaronBallman) [inline clang/test/Sema/return-mismatch.c:8]:
```suggestion
void test1(void) {
```
Same change should be made to all the functions in this file.

File: clang/docs/ReleaseNotes.rst [hunk 9]
```diff
@@ -94,6 +94,7 @@ Deprecated Compiler Flags
 
 Modified Compiler Flags
 -----------------------
+- The `-Wreturn-type` flag has been modified to split some of its functionality into `-Wreturn-mismatch` flag.
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:97]:
```suggestion
- Added a new diagnostic flag ``-Wreturn-mismatch`` which is grouped under
``-Wreturn-type``, and moved some of the diagnostics previously controlled by
``-Wreturn-type`` under this new flag. Fixes #GH72116.
```

File: clang/test/Sema/return-type.c (hunk 5 above)
Comment (11happy) [reply clang/test/Sema/return-type.c]: done

File: clang/test/Sema/return-mismatch.c [hunk 10]
```diff
@@ -0,0 +1,29 @@
+// RUN: %clang_cc1 -Wno-return-type -Wreturn-mismatch -fsyntax-only -verify %s
```
[the remark this answers is not in the document]
Comment (11happy) [reply clang/test/Sema/return-mismatch.c]: done

File: clang/test/Sema/return-mismatch.c [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+// Test that -Wreturn-mismatch is enabled and -Wreturn-type is disabled.
+
+int foo(void) __attribute__((noreturn));
+int bar(void);
+
+void test1() {
+  return 1; // expected-warning{{void function 'test1' should not return a value}}
```
[the remark this answers is not in the document]
Comment (11happy) [reply clang/test/Sema/return-mismatch.c]: done

File: clang/test/Sema/return-mismatch.c (hunk 6 above)
Comment (11happy) [reply clang/test/Sema/return-mismatch.c]: done

File: clang/test/Sema/return-mismatch.c (hunk 7 above)
Comment (11happy) [reply clang/test/Sema/return-mismatch.c]: done

File: clang/test/Sema/return-mismatch.c (hunk 8 above)
Comment (11happy) [reply clang/test/Sema/return-mismatch.c]: done

File: clang/docs/ReleaseNotes.rst (hunk 9 above)
Comment (11happy) [reply clang/docs/ReleaseNotes.rst]: done

**Comment by 11happy** [comment]
Are there any more changes for this PR required from my end?
Thank you

**Review by AaronBallman — APPROVED** [review]
LGTM!
